src/Validation/Common.elm

```diff
@@ -205,65 +217,424 @@ isSubstituable substitution new original =
+            count
+
+
+countForAllQuantifiers : Int -> Formula -> Int
+countForAllQuantifiers count f =
+    case f of
+        Formula.ForAll x subf ->
+            countExistQuantifiers (count + 1) subf
+
+        _ ->
+            count
+
+
+countLeadingQuantifiers : Formula -> Int
+countLeadingQuantifiers f =
+    case f of
+        Formula.ForAll x subf ->
+            countForAllQuantifiers 0 f
+
+        Formula.Exists x subf ->
+            countExistQuantifiers 0 f
+
+        _ ->
+            0
+
+
+removeNQuants : Int -> Formula -> Formula
+removeNQuants n f =
+    if n == 0 then
+        f
+
+    else
+        case f of
+            ForAll _ subf ->
+                removeNQuants (n - 1) subf
+
+            Exists _ subf ->
+                removeNQuants (n - 1) subf
+
+            _ ->
+                f
+
+
+forbiddenVarsErrStr lst =
+    "The variable"
+        ++ (if List.length lst > 1 then
+                "s "
+
+            else
+                " "
+           )
+        ++ "'"
+        ++ String.join "," lst
+        ++ "'"
+        ++ " can't be substituted for"
+
+
+checkSubstitutedVars : List Term -> Term.Substitution -> Zipper.Zipper -> Result (List Problem) Zipper.Zipper
+checkSubstitutedVars allowedVars subst z =
+    let
+        allowed =
+            allowedVars |> List.map Term.toString
+
+        forbidden =
+            subst |> Dict.keys |> List.filter (\v -> not (List.member v allowed))
+    in
+    case forbidden of
+        [] ->
+            Ok z
+
+        lst ->
+            Err <| semanticsProblem z (forbiddenVarsErrStr lst)
+
+
+unaryWithSubstCheck : String -> Signed Formula -> Signed Formula -> Term.Substitution -> Zipper.Zipper -> Result (List Problem) Zipper.Zipper
+unaryWithSubstCheck ruleName refF newF subst z =
+    let
+        referenced =
+            Formula.Signed.getFormula refF
+
+        new =
+            Formula.Signed.getFormula newF
+
+        removedQuantsCount =
+            numOfRemovedQuants referenced new
+
+        allowedVars =
+            varsToBeSubstituted [] removedQuantsCount referenced
+    in
+    checkSubstitutedVars allowedVars subst z
+        |> Result.andThen
+            (\_ ->
+                removeNQuants removedQuantsCount referenced
+                    |> Formula.substitute subst
+                    |> Result.mapError (\str -> semanticsProblem z str)
+                    |> Result.map (\f -> f == new && haveSameSign refF newF)
+                    |> Result.andThen
+                        (\b ->
+                            resultFromBool z
+                                (semanticsProblem z ("Formula was not created by using the " ++ ruleName ++ " rule"))
+                                b
+                        )
+            )
+
+
+notVarsErrStr lst =
+    "The function"
+        ++ (if List.length lst > 1 then
+                "s '"
+
+            else
+                " '"
+           )
+        ++ String.join "," (List.map Term.toString lst)
```

Comment:
`String.join "', '"` – these are multiple terms that occur in the substitution in different places and possibly also in a different order, so each should be in quotes.